File: osgi-framework-bootstrap/src/main/java/net/corda/osgi/framework/JavaAgentLauncher.java
package net.corda.osgi.framework;

import java.io.InputStream;
import java.lang.instrument.Instrumentation;
import java.lang.reflect.Method;
import java.net.URL;
import java.util.Enumeration;
import java.util.Map;
import java.util.Properties;

final class JavaAgentLauncher {

    public static void premain(String agentArguments, Instrumentation instrumentation) throws Exception {
        final ClassLoader cl = JavaAgentLauncher.class.getClassLoader();
        Enumeration<URL> resources = cl.getResources("META-INF/javaAgents.properties");
        while (resources.hasMoreElements()) {
            final Properties properties = new Properties();
            try (InputStream input = resources.nextElement().openStream()) {
                properties.load(input);
            }
            for (Map.Entry<Object, Object> entry : properties.entrySet()) {
                String agentClassName = (String) entry.getKey();
                String agentArgs = (String) entry.getValue();
                Class<?> agentClass = Class.forName(agentClassName, false, cl);
                Method premainMethod = agentClass.getMethod("premain", String.class, Instrumentation.class);
                premainMethod.invoke(null, agentArgs, instrumentation);
            }
        }
    }

    public static void agentmain(String agentArguments, Instrumentation instrumentation) throws Exception {
        premain(agentArguments, instrumentation);
    }
}

File: osgi-framework-bootstrap/src/main/kotlin/net/corda/osgi/framework/JavaAgentLauncher.kt
package net.corda.osgi.framework

import java.lang.instrument.Instrumentation
import java.util.Properties

object JavaAgentLauncher {

    @JvmStatic
    fun premain(@Suppress("UNUSED_PARAMETER") agentArguments : String, instrumentation: Instrumentation) {
        val cl = JavaAgentLauncher::class.java.classLoader
        cl.getResources("META-INF/javaAgents.properties").asSequence().forEach { url ->
            val properties = Properties()
            url.openStream().use(properties::load)
            properties.entries.forEach { entry ->
                val agentClassName = entry.key as String
                val agentArgs = entry.value as String
                val agentClass = Class.forName(agentClassName, false, cl)
                val premainMethod = agentClass.getMethod("premain", String::class.java, Instrumentation::class.java)
                premainMethod.invoke(null, agentArgs, instrumentation)
            }
        }
    }

    @JvmStatic
    fun agentmain(agentArguments : String, instrumentation: Instrumentation) = premain(agentArguments, instrumentation)
}
